Feature/CSL-2007/CSL-2040 group 1 and 2 enhanced tooltips

klown opened the pull request:
Create the "group 2" popover dialogs and define their targets such that they show when the relevant page is loaded, and the criteria for showing them is met.

Note that since this PR is an extension of the group 1 popovers, it is based on and includes the changes in PR #323.  Going forward this PR will track changes for both group 1 and group 2 enhanced tooltips.

JIRAs:
- https://castudl.atlassian.net/browse/CSL-2040
- https://castudl.atlassian.net/browse/CSL-2007

klown committed:
CSL-2007: Migrate tooltips to popover dialogs (first steps)

- add content for 'view' and 'readaloud' popovers
- switch to popover styles and classes

klown committed:
CSL-2007: Append "video" to the video iframe's aria-label

klown committed:
Merge branch 'development' into feature/CSL-2007-group1-enchanced-tooltips

klown committed:
CSL-2007: Convert to popover dialogs for the remaining tips

Converted tooltips to popovers for settings, wordbank, switch,
context, book_actions, and activity.

mbrambilla committed:
CSL-2007: Split apart tooltip and popover template variants, stub out start of chainable tour

mbrambilla committed:
CSL-2007: Fix typo property to variable

klown committed:
CSL-2007: Add `kind` field to `TipType` model

The `kind` field denotes whether the tip is a popup tooltip or a
dialog popover.  The latter are associated with "robust tooltips" and
tours.

klown committed:
CSL-2007:  Remove `kind` field from `TipType` model and fix tour template

- the `kind` of tip -- tooltip vs dialog -- is not used anywhere; remove
  for now.
- fix the name of the settings teacher resource to link to in the
  `popover_tour.html` template.

klown committed:
CSL-2007: Remove migration files as no longer needed

The migrations cancel each other.  The first adds a `kind` field to
the `TipType` model while the second removes it.

klown committed:
CSL-2007: Fix: remove `kind` from display version of `TipType` instances

mbrambilla committed:
CSL-2007: Fix tts issue and settings markup merge error

mbrambilla committed:
CSL-2007- Fix context tip not showing, force focus to popover on show if only one in chain

klown committed:
CSL-2007: Fix name of settings resource url, `SSETTINGS`

mbrambilla committed:
CSL-2007: Fix positioning issue with read aloud popover when using TTS. Stopgap for refocus of control

klown committed:
CSL-2007: Remove empty "Learn more" links from the `TAB` order

klown committed:
CSL-2007: Set focus on the title (header) of the popover dialog when shown

klown committed:
CSL-2007: Revert focus to the `.popover` element when popover dialog is shown

klown committed:
CSL-2040: Add new tip types to fixtures and to logic of TipType.can_show()`

klown committed:
Merge branch 'feature/CSL-2007-group1-enchanced-tooltips' into feature/CSL-2040-group-2-enhanced-tooltips

klown committed:
CSL-2040: Add "student reactions" to popover dialog template

klown committed:
CSL-2040: Add more popover help dialogs from group 2 and adjust code

Added:
- reading_details (Active Reading time and Details),
- reading_data (Student reading data),
- cluey, but only partially (Cluey),
- thoughts (Your thoughts).

With respect to Cluey, the popover dialog is ready, but there is no target
element to attach the popover to.  See comment in CSL-2040:
https://castudl.atlassian.net/browse/CSL-2040?focusedCommentId=36759

klown committed:
CSL-2040: Add group 2 "Manage" popover dialog to the Dashboard

klown committed:
CSL-2040: Add group 2 "Resources" popover dialog to the Dashboard

klown committed:
CSL-2040: Add group 2 "Filters" popover dialog to the Library page

klown committed:
CSL-2040: Add group 2 "Search" popover dialog to the Library page

Also, remove debugging `pdb` statements.

klown commented:
@bgoldowsky , while I've marked this as "ready for review", there are changes to both the kind of tip for Cluey, and its content.  It looks like it will revert to an actual tooltip that is shown on mouse hover, and not a spontaneous display that depends on the history of when it was last shown.  See [Kristen's comment](https://castudl.atlassian.net/browse/CSL-2040?focusedCommentId=36795) in the JIRA.

Given that change, I need to at least change the logic for showing Cluey's popover, likely remove it from that logic.

klown commented:
Based on a conversation with Kristen in the JIRA for CSL-2007, the group 2 [Active Reading time and Details](https://docs.google.com/document/d/1Pod1qiY8agv86j-SMv0pzkdzwYr4KfdjWJ-eDagQNjM/edit#heading=h.6c6bur4n6pdv) is a duplicate of the group 1 [Tooltip_student activity](https://docs.google.com/document/d/1r9sXmIIbpsuMZmpa5iqvw_FoEHSyDxO5YbbSd3VRp1I/edit?usp=sharing).  The latter is a replacement for an existing tooltip and implemented in PR #323.  The former was implemented in this PR as a new popover.  I will remove the new popover from here.  This entails removing a `TipType` from the database and adjusting priorities as appropriate.

The [conversation in the JIRA](https://castudl.atlassian.net/browse/CSL-2007?focusedCommentId=36823) starts with a note that the group 1 popover is missing from the Dashboard tour.

klown committed:
CSL-2040: Remove cluey and student reading details popover dialogs

- `cluey` reverts to a tooltip,
- the `reading_details` popover is a duplicate of the `activity` popover.

klown committed:
CSL-2040: Remove a missed 'reading_details'

Remove 'reading_details' from the list of popovers associated with the
Dashboard.

klown committed:
CSL-2040: Move `resources` popover to a panel on the Dashboard

Also:
- remove check for Cluey from `TipType.can_show()` since it's
not clear where this check belongs; possibly here, or possibly in
client-side javascript.
- special cases for `manage`, `resources`, and `wordbank` popovers
so they are a part of their own page's tooltip tour.

klown committed:
CSL-2040: Handle `resources` and `manage` popovers; fix numerous bugs.

- Handle `resources` and `manage` popovers when they appear on their
  own page,
- In tips `models.py`, regarding the lists of popovers for each page,
  order the list according to tour order.

klown committed:
CSL-2040: Handle `wordbank` popover on the Word bank page

klown committed:
CSL-2040: Fix history updates when a popover help dialog is shown

Add `TV` message (tip viewed) to `messagequeue` when a popover help
dialog is shown.  When the message is POSTed to the server, its
`show_count` and `last_show` time are upated in the database.

klown commented:
@mbrambilla FWIW, after the meeting on Wed afternoon, I put the [page-based lists of the different popover types](https://github.com/cast-org/clusive/blob/cac2e3d909f6e97d9165ee26b6f7f6d4c6ae9925/src/tips/models.py#L21) in tour-order.  That order does not matter for single automatic onboarding display of the popover, but it might be useful to you.

Regarding the meeting itself, look for the section "Tool tip tour questions" in the [meeting notes](https://castudl.atlassian.net/wiki/spaces/CISL/pages/2055176193/2022-10-19+Meeting+notes).

klown committed:
CSL-2040: Moving "Your thoughts" popover from Dashboard to Readings page

klown committed:
CSL-2040: Tidy up `TipType.can_show()`

klown commented:
@mbrambilla By chance, I noticed that when the video in a popover is playing, closing the popover with either the close button, or the "Got it!" button, the audio track of the video keeps playing.  Actually, I think the video itself keeps playing but it is invisible, so it's just heard.

I think that's because closing the popover just hides it: `$curr.CFW_Popover('hide')`, in [tour.js](https://github.com/cast-org/clusive/blob/feature/CSL-2040-group-2-enhanced-tooltips/src/frontend/js/tour.js#L63).  And that's possibly because the popover is part of a tour, and could be shown again by pressing a "previous" button from the next popover in the tour ?  Keeping it around, but hidden, makes the tour easier to present.

By comparison, the previous tooltips were disposed when closed, after first being hidden: `$(this).CFW_Tooltip('dispose');`  from [frontend.js](https://github.com/cast-org/clusive/blob/development/src/frontend/js/frontend.js#L517).

Back to the popovers:  disposing the popover when closed does indeed stop the video and its sound track.  But, I'm not sure it is the right thing to do, or, perhaps the right thing in all contexts.

Alternatively, and possibly a rabbit's hole, use an API for interacting directly with the video and stop it when the popover is closed (hidden)?

mbrambilla commented:
@klown Thank you for the reminder that I needed to look into pausing the YouTube videos on hide of the popovers (and modals).

Will look into possibly hooking into the [YouTube iframe API](https://developers.google.com/youtube/iframe_api_reference), or through more nefarious methods such as resetting the `<iframe>`'s `src` attribute as mentioned in a StackOverflow post about [Youtube Video Still Playing When Bootstrap Modal Closes](https://stackoverflow.com/questions/22613303/youtube-video-still-playing-when-bootstrap-modal-closes)

Also reference for the [SproutVideo Javascript API](https://sproutvideo.com/help/articles/27-javascript_player_api).

klown committed:
CSL-2040: Add triggers for "tip related action" for `resources` tip histories

klown committed:
CSL-2040: Add triggers for`manage` tip histories' tip related actions

klown committed:
CSL-2040: Add triggers for `thoughts` tip histories' tip related actions

klown committed:
CSL-2040:  Add triggers for `filters` tip histories' tip related actions

klown committed:
CSL-2040: Add triggers for `search` tip histories' tip related actions

klown committed:
CSL-2040: Add triggers for `reading_data` tip histories' tip related actions

klown committed:
CSL-2040: Add triggers for `student_reactions` tip histories' tip related actions

klown committed:
CSL-2040: Add trigger for `reading_data` tip related action to "Assigned" and "Popular" tabs on the teacher Dashboard

klown committed:
CSL-2040: Fix misspelling of `data-clusive-tip-action` attribute

klown committed:
CSL-2040: Fix another spelling error re: `data-clusive-tip-action` attribute

klown committed:
CSL-2040: Implement `thoughts` popover for both student Dashboard and Reading pages.

klown committed:
CSL-2040: Implement "Your thoughts" popover for student Dashboard as the `student_reactions` tip type

klown committed:
CSL-2040: Add "My reading" popover for student Dashboard as a `reading_data` tip type

klown committed:
CSL-2040: Add unit test for `TipType.can_shown()`

klown committed:
CSL-2040: Add test case for `switch` when there is only one version of the book

klown commented:
@bgoldowsky @mbrambilla This is ready for review. All the group 1 and group 2 "enhanced tooltips" are in place, including the recent two for the student Dashboard.

The last code I added was a couple of unit tests for changes I made to `TipType.can_show()`.  Good thing, since I found and fixed bug in my changes.  I'm inclined to add more automated tests for the logic underlying what/if a popover should show.  What do you think?

As we discussed at the technical meeting, this PR is ready to merge @mbrambilla's code from PR #333 into it.  Alternatively, this could be reviewed and merged for Q/A while the tour work continues on its branch.  Let me know how you think we should proceed.

klown committed:
CSL-2040: Added test for `TipHistory.register_show()`

klown committed:
CSL-2040: Fix typo in test case

klown committed:
CSL-2040: Add test for `TipHistory.register_show()` for earlier timestamp

When `register_show()` is called with a timestamp that is earlier than
the current history's `last_show`, the value of `last_show` should not
change.

klown committed:
CSL-2040: Change test for `TipHistory.register_show()` to ignore seconds

The relevance of a tip history's `last_show` time's seconds and
microseconds are minor. and are excluded from one test.

klown committed:
CSL-2040: Add test for `TipHistory.register_action()`

bgoldowsky commented on src/library/views.py:
```diff
@@ -389,7 +390,13 @@ def get(self, request, *args, **kwargs):
         context = super().get_context_data(**kwargs)
         context['search_form'] = self.search_form
-        context['tip_name'] = self.tip_shown.name if self.tip_shown else None
+        # BEGIN: Sample Tour
```
Is this a sample, as opposed to something that should remain for production use?

mbrambilla replied:
This is kind of a hybrid mode between the tooltips (tip_name) and the popovers (tours).   This has been cleaned up in PR #336.  I was originally planning on add the tour chains into the original PR, but things forked in two different directions into this PR and then the tour PR.

bgoldowsky commented on src/pages/views.py:
```diff
@@ -223,7 +223,14 @@ def get_context_data(self, **kwargs):
-        context['tip_name'] = self.tip_shown.name if self.tip_shown else None
+        context['clusive_user'] = self.clusive_user
+        # BEGIN: Sample Tour
+        # Sample tour with single item list
```
sample?

mbrambilla replied:
Transitional hybrid mode resulting from trying to do too much in one PR, cleaned up in PR #336.

bgoldowsky commented on src/pages/views.py:
```diff
@@ -710,7 +718,13 @@ def get(self, request, *args, **kwargs):
             'cuelist': json.dumps(cuelist),
             'hide_cues': hide_cues,
-            'tip_name': self.tip_shown.name if self.tip_shown else None,
+            # BEGIN: Sample Tour
```
sample?

mbrambilla replied:
Transitional hybrid mode resulting from trying to do too much in one PR, cleaned up in PR #336.

klown replied:
What are the implications for this PR, then?  Is it simply a matter of removing the "sample tour" comments?  It goes back to the issue of whether the singleton popovers can be merged vs. combining this branch with the tour branch.

klown committed:
CSL-2040: Test for `TipHistory.ready_to_show()`

bgoldowsky commented on src/tips/models.py:
```diff
@@ -24,18 +84,24 @@ class TipType(models.Model):
         """Test whether this tip can be shown on a particular page"""
         # Teacher/parent-only tips
-        if user.role == Roles.STUDENT and self.name in ['book_actions', 'activity']:
+        if user.role == Roles.STUDENT and self.name in TEACHER_ONLY_TIPS:
```
Also Roles.GUEST or use something like "not user.can_manage_periods".  Currently guests are being assigned teacher-only tips.

klown replied:
I should have included a GUEST user in the tests.  When GUEST is tested, the `TipType.can_show()` fails for the `activity` tip on the Dashboard.

By implication, teacher tips should also show for guests using the current version  in production where only `book_actions` and `activity` tips are checked in `TipType.can_show()`.  Do `book_actions` or `activity` tooltip pop-up for guests?  I suspect not because the Dashboard for guests does no include the relevant activity panel nor the book_actions menu.  As such, there is nothing for the tooltip to point to.

But, in the current situation, there are more `TEACHER_ONLY_TIPS` some of whose targets appear for guest users.  So, I'll check for guest users.

bgoldowsky commented on src/tips/models.py:
```diff
@@ -24,18 +84,24 @@ class TipType(models.Model):
         if self.name == 'switch':
             return page == 'Reading' and version_count > 1
+        # Thoughts TipType is only for students
+        if self.name == 'thoughts' and user.role != Roles.STUDENT:
```
Also GUEST, see above.

klown replied:
Right.

bgoldowsky commented on src/shared/templates/shared/partial/popover_tour.html:
```diff
@@ -0,0 +1,384 @@
+        </div>
+        {% elif tour_name == "book_actions" %}
+<!-- book_actions -->
+        {% if clusive_user.role == 'TE' %}
```
This file with all of its 'if' statements feels extremely fragile, it would be great to do some refactoring. Why is this one just TE, while most are TE or PA? Is there some reason that parents should not get to see the video?

klown replied:
Yeah, I'll see if I can refactor and simplify the template.  As for the missing "or PA", that's something like a copy-paste error.

klown replied:
@mbrambilla what kind of grief am I going to cause you due to a refactoring of the `popover_tour.html` template?

mbrambilla replied:
@klown Most of my refactoring in popover_tour.html was around the 'footer' section, as seen here -  you will need to click the 'load diff' link to see the changes: [#336 - popover_tour.html](https://github.com/cast-org/clusive/pull/336/files#diff-eb5c13490b9c5999f2faa4709aacae3eb06246c6dce38c67a9f011f8533a0d3a)

The goal was to create a re-usable template for the footer section to reduce redundant template markup.

klown replied:
Thanks @mbrambilla.  I thought about merging just your footer change into this PR, but it has features that don't apply here or might break things.  I'll work on making a suitable version.  Since that is where most of your refactoring was done, it should make merge conflicts smaller.
🤞

bgoldowsky commented on src/shared/templates/shared/partial/popover_tour.html:
```diff
@@ -0,0 +1,384 @@
+        <p>Click “More actions” (three dots) to assign or customize any book, and for books you uploaded to Clusive, to edit information or delete a book.</p>
+        <div class="row gx-0_5 flex-items-center">
+            <div class="col">
+                {% if has_teacher_resource %}
```
What is the purpose of passing in the 'has_teacher_resource' variable, given that this template has hard-coded the URL of any teacher resource that exists already? In this case, has_teacher_resource is passed in as false by LibraryView. Clearly there is one, but the link is not displayed.

klown replied:
The general rationale of `has_teacher_resource` was flexibility in showing the link.  It should have been named `has_teacher_resource_link` or, even better, `show_teacher_resource_link`.

Admittedly it's somewhat muddled in that the link is not shown for student and guest users, somthing that could be handled by `show_teacher_resource_link`.  But the template covers that situation by the check for a TE or PA role*.

Still, there are cases where the user is a teacher or a parent and the link should not be shown.  For example, if the user has navigated to the Manage page and the Manage popover shows, the link in the popover is odd since it just navigates to the page the user is viewing.  It should not be shown in this case.

As for the Library case, that's a bug; it should not be false.

\*Perhaps this is part of the refactoring:  Move the role check logic to `views.py` to set  `show_teacher_resource_link` and then use only that boolean in the template.  It would remove some clutter.

bgoldowsky commented:
The PR is a little confusing because it contains things related to the tour, but doesn't actually implement the tour. So I think things like the 'tour samples' passed in context variables are harmless, but hope they will be cleaned up in the next PR.

There are defintely some logic errors, mostly due to checking for STUDENT or TEACHER but forgetting about GUEST and PARENT.

klown committed:
CSL-2040: Fix to include guest users when checking if a tip type can show

klown committed:
CSL-2040: Refactor popover tour templates

klown committed:
CSL-2040: Take `ResourceReading` page into account in `TipType.can_show()`

klown committed:
CSL-2040: Fix check for the `resources` tip type on the "Resources" page

klown committed:
CSL-2040: Adjust logic to determine if a "Learn more" link is shown

Some of the logic for showing the "Learn more" link is in the popover
dialog footer template -- if the page requested and the "Learn more"
link are the same, do not show the link.

klown commented:
@boris, @mbrambrilla, I have made changes in response to the review comments, mainly to refactor and reorganize the `popover_tour.html` template.  I think it could be simplified further if some `context` such as the title of the popover and the video link were set in the `views.py`.  But I'm not sure what belongs in the template and what belongs in the code; and am leaning towards the titles being in the template.  On the other hand, if they and other `context` were in the code, then the large `{% if %}` could be eliminated.  Or, perhaps just moved into the `views.py` code.
```
{% if tour_name == ...%} 
... 
{% elif tour_name == ... %}
...
{ % elif ... %}
...
{% endif %}
```

In addition, I noticed another page that was not included in the show-tip logic.  The page is the `ResourcePage`.  The  result is that when using the Reader to view a teacher resource, there are no popovers, and, by implication, there will be no tour.  I modified the code to treat the `ResourcePage` as having the same tip types as the `Reading` page.  Does that make sense?

klown commented:
Regarding more refactoring, I think there is a structure for each popover along the following lines:
```
tour_properties: {
    'settings': {
        'title': 'Settings',
        'video': 'https://videos.sproutvideo.com/embed/a79edbb41414e7c72e/96f3dfacf85e2603?type=hd',
        'tip_text': '<p>Change the font, colors, and more! Settings let you choose the look, supports, and reading tools to make Clusive all your own.</p>',
        'resource_id': 'SSETTINGS',
        'resource_name': 'settings',
    },
    'readaloud': {
        'title': 'Read-aloud',
        'video': 'https://videos.sproutvideo.com/embed/449edbb51c1ae6cecd/90717ea3b61bbdd3?type=hd',
        'tip_text': '<p>Click the page read aloud icon to hear the entire page. Select text and click read aloud in the context menu to hear a smaller selection or a single word read aloud.</p>',
        'resource_id': None,
        'resource_name': 'None,
    },
    ...
}
```
It's a lookup-table where each tip type (e.g., `settings`) defines a set of values for that popover dialog, namely, its title, the link to its video, the tip text, and its teacher resource link, if any.  The properties appropriate to the response are keyed by the `tip_shown` in the `context` and are set as `tour_properties`.

In that case, I _think_ the template's `{%if tour_name == ...%) ... {% elif %} ... {% endif %}` is reduced to just one template where the `tour_properties` fields fill at the appropriate places.

klown commented:
I wrote:
> Regarding more refactoring, I think there is a structure for each popover along the following lines:

I pushed a version of this approach to [another branch](https://github.com/cast-org/clusive/tree/feature/CSL-2040-group-2-enhanced-tooltips-refactor) based on this one.  [This](https://github.com/cast-org/clusive/blob/feature/CSL-2040-group-2-enhanced-tooltips-refactor/src/shared/templates/shared/partial/popover_tour.html#L29) shows the markup for all but one of the popover dialogs.  The `context` popover turns out to be a special case.

klown committed:
CSL-2040: Fix test comparing actual/expected `last_show` times

Allow the test to pass if the actual time is within a minute of the
expected time since, in reality, the actual time is later than the
time passed in when set.
